### PR TITLE
Pauser and renounceAdmin Drips tests

### PR DESCRIPTION
**Pull request by jferas**
This PR contains test for testing manipulation (via governance) of the "pauser" and "renounceAdmin" related functions of the Drips protocol.
Additional PR(s) will test the ability (via governance) to change the "admin" of the Drips protocol.

### CODE REVIEW

**Comment by jferas** [comment]
Current dependencies on/for this PR:
* main
  * **PR #1** <a href="https://app.graphite.dev/github/pr/ScopeLift/radworks-governor-upgrade/1?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #2** <a href="https://app.graphite.dev/github/pr/ScopeLift/radworks-governor-upgrade/2?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #5** <a href="https://app.graphite.dev/github/pr/ScopeLift/radworks-governor-upgrade/5?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #8** <a href="https://app.graphite.dev/github/pr/ScopeLift/radworks-governor-upgrade/8?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #12** <a href="https://app.graphite.dev/github/pr/ScopeLift/radworks-governor-upgrade/12?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #15** <a href="https://app.graphite.dev/github/pr/ScopeLift/radworks-governor-upgrade/15?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #14** <a href="https://app.graphite.dev/github/pr/ScopeLift/radworks-governor-upgrade/14?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **PR #18** <a href="https://app.graphite.dev/github/pr/ScopeLift/radworks-governor-upgrade/18?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
            * **PR #13** <a href="https://app.graphite.dev/github/pr/ScopeLift/radworks-governor-upgrade/13?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ScopeLift/radworks-governor-upgrade/18?utm_source=stack-comment).

File: test/RadworksDripsGovernance.t.sol [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import {IGovernor} from "@openzeppelin/contracts/governance/IGovernor.sol";
+import {ProposalTest} from "test/helpers/ProposalTest.sol";
+
+abstract contract RadworksDripsGovernance is ProposalTest {
+  function setUp() public virtual override(ProposalTest) {
+    ProposalTest.setUp();
+    _upgradeToBravoGovernor();
+  }
+
+  function _proposePassAndExecuteDripsProposal(string memory _description, bytes memory _callData)
```
Comment (alexkeating) [inline test/RadworksDripsGovernance.t.sol:13]: Is this method necessary? It seems like a little less generic version of [`_queueAndVoteAndExecuteProposalWithBravoGovernor`](https://github.com/ScopeLift/radworks-governor-upgrade/blob/3d870db42fd9e4017c595d0d783db6741a8478aa/test/helpers/ProposalTest.sol#L279)

File: test/RadworksDripsGovernance.t.sol [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+    _revokePauserViaGovernance(_newPauser);
+
+    // Ensure the new pauser has subsequently had pauser role revoked
+    assertEq(drips.isPauser(_newPauser), false);
+
+    // Ensure the newly-revoked pauser cannot pause the DRIPS protocol
+    vm.prank(_newPauser);
+    vm.expectRevert("Caller not the admin or a pauser");
+    drips.pause();
```
Comment (alexkeating) [inline test/RadworksDripsGovernance.t.sol:109]: Nit: some of these assertions seem like they could be broken out into separate tests. It could allow us to have more descriptive test names for specific situations. For example, in this last assert we are checking a revert, and to follow the foundry test best practices we would structure our test name to reflect this with a name like `testFuzz_RevertIf_CallerIsNotAPauserOrAdmin`.

Comment (alexkeating) [reply test/RadworksDripsGovernance.t.sol:109]: Also, it would be good to have tests with an existing pauser and admin to be more explicit about other edge cases.

File: src/interfaces/IDrips.sol [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity >=0.8.0;
+
+interface IDrips {
```
Comment (alexkeating) [inline src/interfaces/IDrips.sol:4]: Was this interface generated from `cast` or did we get it from the Radicle source?

File: test/RadworksDripsGovernance.t.sol (hunk 1 above)
Comment (jferas) [reply test/RadworksDripsGovernance.t.sol]: Done: 0d211b97e6805c7d616ed2c0b80d118c3c64809c

File: src/interfaces/IDrips.sol (hunk 3 above)
Comment (jferas) [reply src/interfaces/IDrips.sol:4]: Hand built it from reading Radicle source code.. now using interface generated by `cast interface` command.
Done: e0ed98a23eb6ba0d25c48873324f027b1d48cb97

File: test/RadworksDripsGovernance.t.sol (hunk 2 above)
Comment (jferas) [reply test/RadworksDripsGovernance.t.sol:97]: Done: d5ab4e06fd3aac54151da3b53cb244e77d0a3cfc

File: src/interfaces/IDrips.sol (hunk 3 above)
Comment (jferas) [reply src/interfaces/IDrips.sol:6]: Done: c0ebe7abf0a1a325645990082e24790f671c2159

**Review by alexkeating — APPROVED** [review]

File: test/RadworksDripsGovernance.t.sol [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    );
+    _queueAndVoteAndExecuteProposalWithBravoGovernor(
+      _targets, _values, _calldatas, _description, FOR
+    );
+  }
+
+  function testFuzz_grantPauserOnDrips(address _newPauser) public {
+    assummeNotTimelock(_newPauser);
+    address[] memory _originalPausers = drips.allPausers();
+
```
Comment (wildmolasses) [inline test/RadworksDripsGovernance.t.sol:46]: what if `_newPauser` is already in `_originalPausers`? Think that would cause this test to fail (length would not increment, correct??). Do we need to assume that `!drips.isPauser(_newPauser)`?

File: test/RadworksDripsGovernance.t.sol [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    _grantNewPauserViaGovernance(_newPauser);
```
Comment (wildmolasses) [inline test/RadworksDripsGovernance.t.sol:47]: awesome abstracting of the annoying proposal building 👍

**Review by wildmolasses — COMMENTED** [review]
looks good, mostly nits.

just double checking, is the granting/revoking of pauser (and then the admin renouncement) the only functionality we need to test? if so, looks great.

**Comment by jferas** [comment]
> looks good, mostly nits.
> 
> just double checking, is the granting/revoking of pauser (and then the admin renouncement) the only functionality we need to test? if so, looks great.

There is a separate PR for testing the changing of admin as well as upgrade of the Drips protocol

File: foundry.toml [hunk 6]
```diff
@@ -1,5 +1,5 @@
 [profile.default]
-  evm_version = "paris"
+  evm_version = "shanghai"
```
[the remark this answers is not in the document]
Comment (jferas) [reply foundry.toml:2]: Yes.. only mainnet

**Comment by jferas** [comment]
Added a commit that adds .DS_Store to .gitignore file for smoother MacOS git operations https://github.com/ScopeLift/radworks-governor-upgrade/pull/18/commits/bf585a14b4137012761daa82892af10a402534c4

File: test/RadworksDripsGovernance.t.sol (hunk 4 above)
Comment (jferas) [reply test/RadworksDripsGovernance.t.sol:47]: Indeed.. done: https://github.com/ScopeLift/radworks-governor-upgrade/pull/18/commits/8b89ae610e34dcb41ed910f8823fe60bd52ab907

File: test/RadworksDripsGovernance.t.sol [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    // Ensure the new pauser has been granted pauser role
+    assertEq(drips.isPauser(_newPauser), true);
+
+    // Ensure the the list of pausers got longer by 1
+    assertEq(_originalPausers.length + 1, drips.allPausers().length);
+  }
+
+  function testFuzz_grantedPauserCanPauseAndUnPause(address _newPauser) public {
+    _assumeNotTimelock(_newPauser);
+
```
Comment (wildmolasses) [inline test/RadworksDripsGovernance.t.sol:59]: nit: inconsistent newline
```suggestion
```

**Review by wildmolasses — APPROVED** [review]
i'm scraping the bottom of the barrel on my nits now. think this one's ready for merge

File: test/RadworksDripsGovernance.t.sol (hunk 7 above)
Comment (jferas) [reply test/RadworksDripsGovernance.t.sol:59]: Fixed in a PR farther up the stack, to avoid rebases related to test filename changes: 01e2252faa982cb45cbee6091872ebbb69e2fa90
